rfc-172-graphql-for-govuk.md

```diff
@@ -0,0 +1,509 @@
+|  ->  Index Scan using index_editions_on_document_id_and_state on editions  (cost=0.43..22.29 rows=7 width=769) (actual time=0.003..0.003 rows=1 loops=28)                                                          |
+|        Index Cond: ((document_id = document.id) AND ((state)::text = 'published'::text))                                                                                                                           |
+|Planning Time: 1.115 ms                                                                                                                                                                                             |
+|Execution Time: 0.318 ms                                                                                                                                                                                            |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+Which can be [explored visually in PG Explain][pg-explain-query-plan]
+
+![PG Explain Visualizer showing the execution plan above](rfc-172/pgexplain-query-plan-visualisation.png)
+
+</details>
+
+Still, even with dataloader batching queries and individual queries making good use of indexes, we're asking the database to do at least around 10x more work than the current content-store database has to do, and that will have some performance or infrastructure cost implications. We will need to carefully consider how this solution performs at scale to evaluate these.
+
+Unlike currently, more complicated pages (with lots of links) will be more expensive to render than simple pages. The most complicated page by far on GOV.UK currently is the ministers index page, and in the first prototype GraphQL API this takes between ~500ms and ~1000ms to generate an API response - real world performance may differ, but it will be important to evaluate the performance of the more complicated pages.
+
+In terms of page load times as seen by an end user, there are additional factors that mitigate the impact of API performance. Mainly, the impact of caching - the ~90% of requests which we serve from Fastly's cache won't be affected at all by API performance, while the 10% of cache misses will. We could consider further tuning of our cache behaviour (for example using [stale-while-revalidate][stale-while-revalidate]) to further reduce performance impact.
+
+### Frontend applications will need to make two API calls instead of one
+
+Currently, frontend applications only need to make a single request to content store, and they can then switch on the document type in the response to determine which view to render.
```

Comment:
With a bit of squinting you could describe our current architecture as already having this given we have the call to Router to first look up the content destination and then having the application look up the content. 

So we could, in someways, handle the first query as an aspect of routing which can also determine which code should be used to render an item. At the moment [we have frontend that loads full content items as a means of routing](https://github.com/alphagov/frontend/blob/50d0ffe4db54bbb34398aa4c90536e89e10fb0e5/config/routes.rb#L59-L97), [government frontend that has a single action for _every_ content item](https://github.com/alphagov/government-frontend/blob/30fe1de1026971d08c22a734bf3956dcbfbbb355/config/routes.rb#L26-L29) and [collections that does the routing based on URL structure](https://github.com/alphagov/collections/blob/main/config/routes.rb). If we do further frontend app consolidation we want to standardise on an approach.

Reply:
At one stage I had a whole section in this RFC about how we could have router make the first query, and then pass the schema / document_type through to the frontend as headers. But then I decided it was too much detail / growing the scope too much.

I totally agree that the first query feels like part of routing, and the second one part of rendering. If we're able to consolidate the frontends on a similar timescale to the graphql rollout, it would be great to just do this in a rails routing constraint. Even if we don't do that though, we could explore doing a "which doc type, schema and rendering app should I use for this page" query in router, and then passing these bits through to the frontend. At that point, router could probably be replaced by an off-the-shelf / open source routing component with a custom plugin to do the graphql query ✨